Initial native image agent with JVM mode tests integration

galderz opened the pull request:
Prototype implementation for #36822

Here are the implementation details:

It adds a new profile, called `native-with-agent` to the generated projects where the JVM mode tests run with the native image agent. This profile sets the native image agent parameter necessary to run the JVM mode tests with. This parameter includes references to filters that are produced before running JVM mode tests.

Before the JVM mode tests run, `GenerateCodeTestsMojo` has been modified to generated the native image agent filters are produced. This enables filtering out most of the configuration that is quarkus specific, but crucially it doesn't handle resources. One caveat of this early filtering is that we don't have context of the application to do more fine grained filtering. An alternative implementation would skip filtering at this level and instead filter at the build step phase, just like resources are filtered (read on).

During JVM mode tests, native image agent generates config files inside `target/generated-native-config`.

Then, native image agent specific build steps pick up this configuration, parse it, filtering using regex and produce existing build items, e.g. `ReflectiveClassBuildItem`, `NativeImageResourcePatternsBuildItem`...etc.

There is a builder class that produces json but there's no class to read json files, so I've gone ahead and created a small, dependency-free, json parser called `JsonReader` for this. The objective of this parser was to get something working quickly and without dependencies, so I'm sure there's a lot of ways in which it can be improved :)

As mentioned above, resources cannot be filtered at the native image agent level, so I post-processed the resources configuration file using the `JsonReader` parser, along with regex, to filter out resources that Quarkus takes care of.

One important note here is that both the native image configuration that Quarkus produces, and the one extracted from the native image agent ends up in the same configuration files inside the generated bytecode. Achieving this is quite simple because we just reuse the same existing build items. But it might make debugging easier if the files were different so that we can distinguish between those Quarkus produces and those that come from the native image agent more easily. Not sure this is doable, @zakkak? Otherwise, you can always run the app with `-Dnative` and `-Dnative-with-agent` and compare the json files (see a demonstration below).

I've tested out this in a new project generated from this branch called `new-project` (attached [new-project-0.1.tar.gz](https://github.com/quarkusio/quarkus/files/13236192/new-project-0.1.tar.gz)). This is a small example that requires reflection to be configured for `Alice` class lookup in order to respond to the `/hello/Alice` endpoint. The native image agent integration works when the native integration test for that endpoint is a success. Comparing the output of the following commands with `-Dnative` and `-Dnative-with-agent` you can see what impact the native image agent jvm mode test integration has on the configuration:

```
$ ./mvnw verify -Dnative 
...
$ tar xfO ./target/quarkus-app/quarkus/generated-bytecode.jar META-INF/native-image/reflect-config.json | jq -s '.[] |= sort_by(.name)' > target/reflect-config-no-agent.json
...
$ ./mvnw verify -Dnative-with-agent
...
$ tar xfO ./target/quarkus-app/quarkus/generated-bytecode.jar META-INF/native-image/reflect-config.json | jq -s '.[] |= sort_by(.name)' > target/reflect-config-with-agent.json
...
$ diff target/reflect-config-with-agent.json target/reflect-config-no-agent.json
619,628d618
<     },
<     {
<       "methods": [
<         {
<           "parameterTypes": [],
<           "name": "sayMyName"
<         }
<       ],
<       "allDeclaredConstructors": true,
<       "name": "org.acme.Alice"
632,635d621
<       "name": "org.acme.Bob"
<     },
<     {
<       "allDeclaredConstructors": true,
```

It has added `Alice` class and its `sayMyName` method to the reflection configuration, which is required in order for endpoint to respond.

There are also differences in the resource configuration but they are not relevant here:

```
$ tar xfO ./new-project/target/quarkus-app/quarkus/generated-bytecode.jar META-INF/native-image/resource-config.json | jq -s '.[].resources.includes |= sort_by(.pattern)' ...
...
$ diff with-agent/resource-config-with-agent.json no-agent/resource-config-no-agent.json
31,78d30
<         },
<         {
<           "pattern": "\\\\QMETA-INF/maven/org\\.apache\\.maven/maven-core/pom\\.properties\\\\E"
<         },
<         {
<           "pattern": "\\\\QMETA-INF/plexus/components\\.xml\\\\E"
<         },
<         {
<           "pattern": "\\\\QMETA-INF/services/java\\.lang\\.System\\$LoggerFinder\\\\E"
<         },
<         {
<           "pattern": "\\\\QMETA-INF/services/java\\.net\\.spi\\.InetAddressResolverProvider\\\\E"
<         },
<         {
<           "pattern": "\\\\QMETA-INF/services/java\\.net\\.spi\\.URLStreamHandlerProvider\\\\E"
<         },
<         {
<           "pattern": "\\\\QMETA-INF/services/java\\.nio\\.channels\\.spi\\.SelectorProvider\\\\E"
<         },
<         {
<           "pattern": "\\\\QMETA-INF/services/java\\.nio\\.file\\.spi\\.FileSystemProvider\\\\E"
<         },
<         {
<           "pattern": "\\\\QMETA-INF/services/java\\.time\\.zone\\.ZoneRulesProvider\\\\E"
<         },
<         {
<           "pattern": "\\\\QMETA-INF/services/javax\\.xml\\.stream\\.XMLInputFactory\\\\E"
<         },
<         {
<           "pattern": "\\\\QMETA-INF/services/org\\.codehaus\\.groovy\\.runtime\\.ExtensionModule\\\\E"
<         },
<         {
<           "pattern": "\\\\QMETA-INF/sisu/javax\\.inject\\.Named\\\\E"
<         },
<         {
<           "pattern": "\\\\Qjava/io/Serializable\\.class\\\\E"
<         },
<         {
<           "pattern": "\\\\Qorg/apache/maven/model/pom-4\\.0\\.0\\.xml\\\\E"
<         },
<         {
<           "pattern": "\\\\Qorg/osgi/annotation/bundle/Requirements\\.class\\\\E"
<         },
<         {
<           "pattern": "java\\.base:\\\\Qjava/io/Serializable\\.class\\\\E"
<         },
<         {
<           "pattern": "java\\.base:\\\\Qjdk/internal/icu/impl/data/icudt72b/nfc\\.nrm\\\\E"
```

One thing of interest in the resource differences is that it uncovers a bug in the native image integration. It's escaping patterns when it shouldn't, see #36823.

Final note, what to filter for is up for debate. I was quite aggressive in the native image agent filters because there really wasn't need for any of the additional configuration that it generated, but certain filters are debatable. Not sure what is the best approach here. Do we prefer minimal filtering that we now for sure Quarkus takes care of? Or do we prefer aggressive filtering? I think I would probably go for minimal because the aim here is to get something that native image builds and works. The user can trim things later if necessary.

maxandersen commented on core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAgentResourceConfigStep.java:
```diff
@@ -0,0 +1,48 @@
+                .resolve(Path.of("generated-native-config", "resource-config.json"));
+        if (agentResourceConfigPath.toFile().exists()) {
+            final Pattern ignoreResources = Pattern.compile(
+                    "^((?!(microprofile|quarkus|smallrye|jboss|junit|slf4j|application.properties|surefire|logging.properties|Test.class|java/lang|java/util)).)*$");
```
what makes something qualified /necessary to have in this list?

galderz replied:
Good question.

I created the list based on the output of the native image agent config for the getting started reactive sample JVM mode unit tests, see [here](https://gist.github.com/galderz/2008721d5b04f31bcc40cc67eb0b71e9).

Something more dynamic would likely be needed here. I guess there could be some general starting point for this list, and then based on the extensions in use, expand it?

maxandersen commented on core/builder/src/main/java/io/quarkus/builder/JsonReader.java:
```diff
@@ -0,0 +1,442 @@
+ * A json format reader.
+ * It follows the <a href="https://www.json.org/json-en.html>ECMA-404 The JSON Data Interchange Standard.</a>.
+ */
+public class JsonReader {
```
I just know that this class is going to be misused to add more json stuff....

galderz replied:
That can of worms was open when the `Json` class was added to generate build output metrics, which Foivos' then reused to generate the json config output for native image :)

maxandersen commented on independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/buildtool/maven/base/pom.tpl.qute.xml:
```diff
@@ -206,6 +206,33 @@
+                <skipITs>false</skipITs>
+            </properties>
+            <build>
+                <plugins>
```
this is just for tests....but how about 'quarkus run' / `mvn quarkus:run` - should that be configured somehow?

galderz replied:
Yeah that should be done too. I'll look into it.

galderz replied:
@maxandersen I've looked into add quarkus run integration but I should be deferred to a later stage (we can create a follow up issue). Gradle `quarkusRun` does not seem to work at all. It starts Quarkus and then finishes the build, see [here](https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/Gradle.20quarkusRun.2C.20does.20it.20work.20at.20all.3F). Also, there are no integration tests under `gradle` that verify `quarkusRun` works at all. Native image agent integration with run targets should be added after that is fixed.

zakkak commented:
> But it might make debugging easier if the files were different so that we can distinguish between those Quarkus produces and those that come from the native image agent more easily. Not sure this is doable, @zakkak? Otherwise, you can always run the app with -Dnative and -Dnative-with-agent and compare the json files (see a demonstration below).

@galderz that's definitely possible, but we would need to explicitly pass these config files to native-image using `-H:ReflectionConfigurationFiles`, `-H:ResourceConfigurationFiles`, `-H:JNIConfigurationFiles`, and `-H:SerializationConfigurationFiles` which are considered "experimental". 

I wonder if we could add some kind of comment in the build items and output it in the json files, in the future we could augment [the proposed `reasons` field](https://github.com/oracle/graal/issues/7482).

Either way would be acceptable I believe.

zakkak commented:
> Do we prefer minimal filtering that we now for sure Quarkus takes care of? Or do we prefer aggressive filtering? I think I would probably go for minimal because the aim here is to get something that native image builds and works. The user can trim things later if necessary.

Minimal sounds good, as long as the user can include and edit the generated parts in their source code repository. Things might become tricky though when later you do an update and rerun `native-with-agent` since you will have to do the clean up again.

zakkak commented on core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAgentReflectConfigStep.java:
```diff
@@ -0,0 +1,63 @@
+                .map(JsonReader.JsonString::value)
+                .toArray(String[]::new);
+        reflectMethodProducer.produce(
+                new ReflectiveMethodBuildItem(className, method.<JsonReader.JsonString> get("name").value(), parameterTypes));
```
Why the `json -> BuildItem -> json` step? Why not just merge the json files?

galderz replied:
Hmmm, care to share more details? I don't see how it would be vastly different or better than this.

zakkak commented on devtools/maven/src/main/java/io/quarkus/maven/GenerateCodeTestsMojo.java:
```diff
@@ -12,5 +19,123 @@ public class GenerateCodeTestsMojo extends GenerateCodeMojo {
+            Files.write(path, content.getBytes(StandardCharsets.UTF_8));
+        } catch (IOException e) {
+            throw new MojoExecutionException("Unable to write native image agent access filter file", e);
+        }
```
Ultimately we will need the corresponding code for gradle.

zakkak reviewed (commented):
Very interesting and less complicated than I expected. Nice work @galderz.

maxandersen commented:
don't we control the exeuction of tests and run enough to run with different agents? feel like i recall @stuartwdouglas or @geoand mention we could register agents?

that could avoid need of manipulating pom.xml/build.gradle?

galderz commented:
Thanks all for the feedback. I've had a quick chat with @zakkak to clarify a few things. I'll work on the next iteration of the feature and I'll post an update when ready.

zakkak commented on core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAgentResourceConfigStep.java:
```diff
@@ -0,0 +1,48 @@
+                stream.<JsonReader.JsonString> map(include -> include.get("pattern"))
+                        .map(JsonReader.JsonString::value)
+                        .filter(pattern -> ignoreResources.matcher(pattern).find())
+                        .forEach(patternBuilder::includeGlob);
```
@galderz I think the following resolves https://github.com/quarkusio/quarkus/issues/36823

```suggestion
                        .forEach(patternBuilder::includePattern);
```

galderz replied:
Ah ok. I'm going with a diffeerent approach where I won't be using this anyway. I'll close #36823.

galderz commented:
I've just pushed a new prototype that works like this:
* The maven project generation still adds a new profile `native-with-agent` and that makes surefire unit tests run with native image agent.
* The native image agent spits out the configuration into `target/native-agent-base-config` folder for maven.
* There is no longer filtering done at the agent level and instead all the configuration filtering is delegated to the quarkus build inside `NativeImageAgentConfigStep`.
* `NativeImageAgentConfigStep` reads the native image agent json using the new `JsonReader` class and that is then transformed using the new `JsonTransform` interface, combined with the existing `Json` class that emits json (The transformation API is inspired by how the [upcoming Class-File API](https://openjdk.org/jeps/457) in OpenJDK works). So, there are no longer intermediate build items created. It just takes json in, transforms it and spits it out onto the folder `<source-jar-dir>/native-agent-config` folder for maven.
* The `JsonTransform` API only has one function, which is to drop json from the input. That's all that is needed for this use case, but more functions could be added for future use cases.
* The `Json` API has been enhanced to add a new `skipEscape` option to avoid escaping strings where they are already escape. That's the case with transformation where the input json is already escaped and hence does not need to be escaped again.
* I've created different regular expression filters for each configuration file. The regular expressions are simple: if there is a reflection class FQN containing the word `quarkus`, it's filtered out...etc. For each configuration file type, there's a different lambda that figures out at which level the filtering needs to happen.
* The contents of `native-agent-config` are hooked into native image via `-H:ConfigurationFileDirectories=` option. @zakkak Looking at the source code this option is stable (iow not experimental), so seems like a better option that individual file configuration parameters.
* `integration-tests/maven/src/test/java/io/quarkus/maven/it/NativeAgentIT` has been added that verifies that native image integration unit tests work for a project that needs reflection configuration to succeed. See associated project code. 
* `integration-tests/gradle/src/test/java/io/quarkus/gradle/nativeimage/NativeAgentIntegrationTestIT.java` has been added that verifies the same thing but gradle projects.

I have the following questions, doubts or comments:
* The maven project unit test bakes the new profile that would be seen by new projects generated which is not ideal, but I don't see tests that generate maven projects on the file. Any thoughts on this?
* The Gradle project is in `integration-tests/gradle/src/main/resources/it-test-native-agent-project` and there I manually added the test `jvmArgs` to contain the native image agent parameters (note that the folder where it spits it out is instead `build/native-agent-base-config`). I do see gradle project generation is in quarkus, but how would you add a profile (ala -Dnative-with-agent in maven) in Gradle so that those `jvmArgs` are only added when the given profile/option is active?
* Seems to me the generic type in `Json.JsonBuilder` adds unnecessary complexity. Its `self` method, which is the only one that uses the generic is not used at all.
* It probably makes sense to have a single regular expression that works for all configuration file types, because in the end what is filtered is common. This avoids duplication of keywords to match. It is also true that sometimes you're filtering FQN classes and other times resource paths and so the point could be made that finer tuned regular expressions would result in different regex for different configuration files, but seems to me that would makes regex overcomplicated.
* I explored adding `quarkus:run` integration for maven and that might be doable, but the equivalent feature for Gradle needs further work (see https://github.com/quarkusio/quarkus/issues/37379), so I think it's better to defer that to a later stage.

vsevel commented:
when you are generating the missing declarations, it would be nice to display it somewhere readable. for instance in the build logs.
this would allow the developer to figure out what he missed, and give him the option to add the declaration explicitly in the app.

zakkak commented:
Great work, @galderz. I like the new approach even better :)

vsevel commented:
> I could add some log messages that specify the folders where the native image agent generated config files and the transformed ones are located.

ideally I would give more control to the developer. at least in our env, dev teams do not have access to the CI workspace. they have to go through the CI team to get an extract.
I would rather print a log message saying that 26 classes have been added to the reflection json file, and offer to specify an option `-DdisplayNativeAgentFileContent=true` to show the content. unless you think this could be huge (we could even add a quality gate searching for this pattern). 
ideally I would like the dev team to be autonomous. I suspect the CI team is going to be bored pretty quickly. and I suspect they will not want to provide access to the nodes.

galderz commented:
> > I could add some log messages that specify the folders where the native image agent generated config files and the transformed ones are located.
> 
> ideally I would give more control to the developer. at least in our env, dev teams do not have access to the CI workspace. they have to go through the CI team to get an extract. I would rather print a log message saying that 26 classes have been added to the reflection json file, and offer to specify an option `-DdisplayNativeAgentFileContent=true` to show the content. unless you think this could be huge (we could even add a quality gate searching for this pattern). ideally I would like the dev team to be autonomous. I suspect the CI team is going to be bored pretty quickly. and I suspect they will not want to provide access to the nodes.

Ok, I understand the situation now. I think an option to show native image agent configuration files in the console would make sense to cover such scenario, but it would work best if only the post-processed configuration would be shown. So, not what comes out directly out of the native image agent, but after having that filtered out to remove things that Quarkus takes care of.

For the most basic of Quarkus applications that has a just returns a greeting message, these are the sizes of files generated by native image agent:

```bash
$ l ./new-project/target/native-agent-base-config
total 872
drwxr-xr-x   9 galder  staff   288B  7 Dec 09:45 .
drwxr-xr-x  18 galder  staff   576B  7 Dec 09:46 ..
-rw-r--r--   1 galder  staff   967B  7 Dec 09:45 jni-config.json
-rw-r--r--   1 galder  staff    86B  7 Dec 09:45 proxy-config.json
-rw-r--r--   1 galder  staff   406K  7 Dec 09:45 reflect-config.json
-rw-r--r--   1 galder  staff   8.6K  7 Dec 09:45 resource-config.json
-rw-r--r--   1 galder  staff   2.1K  7 Dec 09:45 serialization-config.json
```

The biggest file there is by quite a big marging `reflect-config.json` with 406K. Once post-processed I see:

```
$ l ./new-project/target/new-project-1.0.0-SNAPSHOT-native-image-source-jar/native-agent-config
total 152
drwxr-xr-x  7 galder  staff   224B  7 Dec 09:45 .
drwxr-xr-x  6 galder  staff   192B  7 Dec 09:46 ..
-rw-r--r--  1 galder  staff   785B  7 Dec 09:45 jni-config.json
-rw-r--r--  1 galder  staff     2B  7 Dec 09:45 proxy-config.json
-rw-r--r--  1 galder  staff    55K  7 Dec 09:45 reflect-config.json
-rw-r--r--  1 galder  staff   4.8K  7 Dec 09:45 resource-config.json
-rw-r--r--  1 galder  staff   767B  7 Dec 09:45 serialization-config.json
```

The reflection configuration has dropped to 55K. Technically this could be further brought down but it's not easy for Quarkus to know exactly, because these remaining registrations are all related to `java`, `javax` and `sun` packages, so the application itself could have caused those to be required. So, the post-processing is quite conservative here, but if printed out they can always be added to the source code and do further filtering there if needed.

vsevel commented:
> So, not what comes out directly out of the native image agent, but after having that filtered out to remove things that Quarkus takes care of.

yes

> but if printed out they can always be added to the source code and do further filtering there if needed.

yes exactly
having the agent to detect something means that the developer has forgotten declaring some resources in the app. and we are down to having the tests to identify the missing pieces. if a test is disabled at some point, we could forget declaring some resources, and get a regression on the native build.
it is kind of awkward if you think about it: activating or deactivating a test now has an impact on the built app, whereas tests should only observe and report.

vsevel commented:
if/when your reporting feature is available, I would instruct the teams to review the generated content systematically, as it is smell to detect something. I think this should be logged as a WARN

galderz commented:
I had a follow up chat with @aloubyansky last week and I'm working on a further iteration of this. I'm also including the feedback that @vsevel provided. I will update once I have something to show once again.

galderz commented:
I've just pushed some updates, here are the relevant changes:

* The native image agent integration impact is now informative rather than affecting the native executable build by default. This addresses potential issues raised by @vsevel whereby ignoring tests could lead to different native image outcomes. An example of the output would look something like this:

```
[INFO] [io.quarkus.deployment.steps.NativeImageAgentConfigStep] Discovered native image agent generated files
[INFO] [io.quarkus.deployment.steps.NativeImageAgentConfigStep] Generated reflect-config.json:
[
{
  "name":"org.acme.Alice",
  "methods":[{"name":"<init>","parameterTypes":[] }, {"name":"sayMyName","parameterTypes":[] }]
},
{
  "name":"org.acme.Bob"
}
]
```

This output currently appears as part of the native image process. I could potentially print this once unit tests has completed (it would probably a better time for it), but couldn't see how to adjust a mojo to do such a thing. @aloubyansky any idea?

* I added an option to the native agent called `agentConfigurationApply`. This option is disabled by default, but when enabled it applies the native image agent generated and post-processed configuration into the current native executable process. This can be used by users to skip the step of copying the generated configuration that is printed in the log to separate files and you can use it right away. However, this should be done with caution for the reasons explained above. I'll expand the javadoc option with some warning advice.

* I've brought back generation of access and caller json filters for the native image agent. With the guidance of @aloubyansky I've added this to a new `GenerateNativeImageAgentFiltersMojo` class that consumes the application model to construct the filters. The class adds filters to skip packages located in dependencies that either test scope, or deployment classpath only (IOW, not in runtime classpath). It then takes all these packages to skip and shrinks them to 3 component packages (e.g. `a.b.c`). The number 3 is a bit arbitrary but seemed to be enough to filter undesired packages for now. We could up this to say 4/5 package components if needed.

* Resource configuration is not handled by the native image agent access or caller filters, so post-processing for this file remains. This filtering here is still based on hardcoded regular expressions, but I will try to expand this filtering further to skip resources in test or deployment classpath only.

* Filtering at the native image level level and post-processing for resource configuration has very positive impact reducing the configuration, but some very small outlayers remain due to caller information not present and since they are all related to JDK classes, I don't think I can filter those out 100%. Full output of a integration test added in this PR can be found [here](https://gist.github.com/galderz/8f1782b4a49fb75ddc190e07fe11d016).

* Note that the configuration output is nicely formatted for all but resource configuration. This is due to post-processing happening for resource configuration, which is handled by the JSON generator in the builder project. I'll look into adding some formatting there.

aloubyansky commented on devtools/maven/src/main/java/io/quarkus/maven/GenerateNativeImageAgentFiltersMojo.java:
```diff
@@ -0,0 +1,151 @@
+import io.quarkus.runtime.LaunchMode;
+
+@Mojo(name = "generate-filters", defaultPhase = LifecyclePhase.PROCESS_TEST_SOURCES, requiresDependencyResolution = ResolutionScope.COMPILE_PLUS_RUNTIME, threadSafe = true)
+public class GenerateNativeImageAgentFiltersMojo extends GenerateCodeMojo {
```
It should probably extend `QuarkusBootstrapMojo` instead.

aloubyansky commented on independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/buildtool/maven/base/pom.tpl.qute.xml:
```diff
@@ -140,6 +140,7 @@
                             <goal>build</goal>
                             <goal>generate-code</goal>
                             <goal>generate-code-tests</goal>
+                            <goal>generate-filters</goal>
```
If this goal is relevant only in the `native-with-agent` profile then it could be configured only in that profile.

aloubyansky commented:
> I could potentially print this once unit tests has completed (it would probably a better time for it), but couldn't see how to adjust a mojo to do such a thing. @aloubyansky any idea?

Sounds like it would have to be another goal that would be executed after the tests.

galderz commented:
@aloubyansky @zakkak @maxandersen @vsevel This PR is not a draft any more. I've updated the description of #36822, which explains in detail the motivation, how the integration works and how to debug it. Parts of it will become the documentation once consensus is achieved.

vsevel commented:
> @aloubyansky @zakkak @maxandersen @vsevel This PR is not a draft any more. I've updated the description of #36822, which explains in detail the motivation, how the integration works and how to debug it. Parts of it will become the documentation once consensus is achieved.

Hi @galderz that is my attention to spend some time on it

galderz commented:
Looking into the test failures.

zakkak commented on core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:
```diff
@@ -974,6 +984,9 @@ public NativeImageInvokerInfo build() {
                 }
 
+                nativeImageAgentConfigDirectory
+                        .ifPresent(dir -> nativeImageArgs.add("-H:ConfigurationFileDirectories=" + dir.getDirectory()));
```
This is overriding the default, but that's OK since we merge the native-image-agent configurations with the ones generated by Quarkus, so everything will end up under this directory, right?

galderz replied:
This was a very interesting comment. The bottom line is that the line I've added doesn't override the default. It complements it. IOW, any `META-INF/native-image/*.json` the native image build generates will carry through, as well as anything inside the `-H:ConfigurationFileDirectories` above.

You can see this in action (I will adjust the test to verify this) by doing the following. I added `Carol` to the native image agent IT in this PR that looks like this:

```java
@RegisterForReflection
public class Carol
{
    public String sayMyName()
    {
        return "Carol";
    }
}
```

Then I made sure that the unit test did not exercise this class, so the native image agent reflection configuration was:

```json
[
{
  "name":"org.acme.Alice",
  "methods":[{"name":"<init>","parameterTypes":[] }, {"name":"sayMyName","parameterTypes":[] }]
},
{
  "name":"org.acme.Bob"
}
]
```

Through annotation processing, `Carol` gets added to `META-INF/native-image/reflect-config.json`:

```
  {
    "name": "org.acme.Carol",
    "allDeclaredConstructors": true,
    "allDeclaredMethods": true,
    "allDeclaredFields": true
  },
```

Finally, I run the native executable and I was able to make this call successfully:

```
❯ curl http://localhost:8080/hello/Carol
Hello Carol%
```

This means that the registrations generated by Quarkus are complementary.

zakkak reviewed (changes requested):
Hi @galderz, I went through it adding some comments. Please have a look. 

Note: naming and refactoring suggestions are subjective so I understand if we end up not agreeing

galderz commented on .github/workflows/ci-actions-incremental.yml:
```diff
@@ -1041,7 +1041,7 @@ jobs:
         id: setup-graalvm
         uses: graalvm/setup-graalvm@v1
-        if: startsWith(matrix.os-name, 'windows')
+        if: "startsWith(matrix.os-name, 'windows') || startsWith(matrix.category, 'Misc1')"
```
> Judging by the commit message in [c3e1d75](https://github.com/quarkusio/quarkus/commit/c3e1d75a675839754ddd244f0106b3e43d3b4eaf), in order for the integration to work one needs to have Mandrel or GraalVM locally installed which is non-optimal as the default is to use the builder-image (and saves people from a lot of trouble).
> 
> That seems OK for the time being, but IMO we should make it work with the builder image as well.

I'm not sure about the real benefit of getting the builder image `JAVA_HOME` to be used to run the surefire unit tests. It can probably be done but I have doubts about doing it:

* First, the way I describe the feature (see https://github.com/quarkusio/quarkus/issues/36822), this feature would be used by a subset of Quarkus users, those that have the need to bring a library/component that requires some native image configuration, or those that use JDK APIs that require extensive native image configuration. Since it doesn't affect all Quarkus users, I think it's a small price to pay to ask those to install GraalVM locally and run the `native-with-agent` profile with their unit tests.

* Secondly, I see this feature as something that gives you a head start with the required native image configuration. So you would run it once at the start, get the configuration files, add them to your source control and extend as needed. Maybe you would run it again once you have made substantial changes or upgraded to a newer version. The point I'm trying to make here is that you would not be running this profile very often.

So, combining the two points above: the fact that it only affects a subset of Quarkus users and this is something that you would only run once in a while, I'm not sure of its benefits. I agree that it could be useful to remove the need for a local installation, and maybe it's quite an easy thing to do, but right now I'm not convinced.

> In any case, I would prefer the native-image agent integration test to be moved to a separate IT category and only use a local installation when testing it, to keep all testing consistent except for the actual cases that need to run in a different setup.

Yeah that makes sense. I went for a quick solution here to see if the CI configuration change would work, but I agree that such integration test should be moved to its own category. I'll try to do that within this PR.

galderz commented on core/builder/src/main/java/io/quarkus/builder/Json.java:
```diff
@@ -26,6 +26,10 @@ public final class Json {
     private static final int CONTROL_CHAR_START = 0;
     private static final int CONTROL_CHAR_END = 0x1f;
     private static final char CHAR_QUOTATION_MARK = '"';
+    private static final String DEFAULT_INDENT = "  ";
+    private static final String EMPTY_STRING = "";
+    private static final String NEW_LINE = System.lineSeparator();
+    private static final String ENTRY_SEPARATOR_NEW_LINE = "," + NEW_LINE;
```
See @vsevel's https://github.com/quarkusio/quarkus/pull/36826#issuecomment-1839160473 comment for motivation for displaying the generated configuration.

I can see the benefit of displaying the configuration as part of the build process, as well as emitting actual files. I agree that it's a pain having to add the pretty printing logic. The best alternative I can think for anyone that wants to display the generated configuration file without the pretty printing logic would to configure CI such that:

1. Runs unit tests with `native-with-agent` (e.g. `mvn test`)
2. Install `jq` or similar
3. Invoke `jq` to pretty print the json files.
4. Run further build steps without re-running unit tests.

Steps 2/3 could be moved to a latter stage depending on the complexity of doing step 4. @vsevel how does this sound?

Note about increasing the configuration size. The configuration files produced by the native image agent are generated pretty printed. What I have added is pretty printing for the json files that are post-transformed, which is the resource configuration json file. I did that because the JSON that comes out of the builder JSON printed is not formatted.

zakkak replied:
Oh, I forgot about @vsevel's request. I leave this up to you two. I am OK with pretty printing, I was just concerned about the increased complexity.

galderz commented on core/builder/src/main/java/io/quarkus/builder/JsonTransform.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.function.Predicate;
+
+@FunctionalInterface
+public interface JsonTransform {
```
I think `Transform` is the better name. That's the name of the function (as annotated by the `@FunctionalInterface`), it transforms. Btw, the transform part is heavily inspired by the upcoming [Class-File API](https://openjdk.org/jeps/457) where they use `Transform` instead of `Transformer`.

galderz commented on core/builder/src/main/java/io/quarkus/builder/JsonTransform.java:
```diff
@@ -0,0 +1,15 @@
+
+@FunctionalInterface
+public interface JsonTransform {
+    void accept(Json.JsonBuilder<?> builder, JsonReader.JsonValue element);
```
Again naming in line with Class-File API, but I'm curious, why the name `keep`?

zakkak replied:
Because of `dropping`. One of the keeps and the other one drops, but you essentially override `keep` in `dropping` so it doesn't make sense. +1 on following the Class-File API. Sorry for the noise.

galderz replied:
Right, `dropping` is just one type of transformation you can implement with this. IOW, `dropping` implements `accept` to drop certain JSON. For this particular use case, it's the only high level transformation I required, but you could other ones that transform different parts of the JSON.

galderz commented on core/builder/src/main/java/io/quarkus/builder/Json.java:
```diff
@@ -354,4 +494,21 @@ static String escape(String value) {
+
+        private ResolvedTransform(Json.JsonBuilder<?> resolvedBuilder, JsonTransform transform) {
+            this.resolvedBuilder = resolvedBuilder;
+            this.resolvedBuilder.setTransform(transform);
```
`ResolvedTransform` binds together the transform function and the builder. It allows to "ship" the builder to the user-provided transform function via this internal binding transform class.

Your suggestion won't work alone. You need that some way to pass the builder onto the JSON value for transformation. If not using the `ResolvedTransform`, you can change `forEach(JsonTransform transform)` to be `forEach(JsonTransform transform, Json.JsonBuilder<?> builder)` instead and then do:

```
        @Override
        public void forEach(JsonTransform transform, Json.JsonBuilder<?> builder) {
            members().forEach(member -> transform.accept(builder, member));
        }
```

I think the `ResolvedTransform` option is cleaner from an API perspective, but it's a bit more complex from an implementation perspective.

vsevel commented:
> I think it's a small price to pay to ask those to install GraalVM locally and run the native-with-agent profile with their unit tests.

I agree with @zakkak . I wish we did not have to install GraalVM on our build machines. today we do not. we are building the native image using the build image. and that saves a lot of pain, considering that many quarkus applications are using different versions of quarkus, hence different versions of graalvm/mandrel. 
at this time we target native first for all of our applications that do not use an excessive amount of memory, and/or do not the absolute best performances. that means that almost all of our quarkus applications target mandrel today. having this requirement would mean we have to install every single version of mandrel, and make sure we are using the right one for a given quarkus version. feasible, but I can see the eyes of the soft factory lead when I will be telling him that ;-)
for now, I can provide some feedback on a special host I am using for my native tests, which has versions of graalvm installed, but going forward if this requirement stays, the feature is probably going to be used lightly.

galderz commented:
> I agree with @zakkak . I wish we did not have to install GraalVM on our build machines. today we do not. we are building the native image using the build image. and that saves a lot of pain, considering that many quarkus applications are using different versions of quarkus, hence different versions of graalvm/mandrel.

Fair enough. I'll explore this option to see how JVM mode unit tests can run with the GraalVM installation within the builder image.

maxandersen commented:
If I understand it right I would say we should fix or at least make it easier to match the right graalvm/mandrel with quarkus than keep doing things that is only possible with docker. 

Remember that the native agent will potentially record different things dependent on which OS it is running on.

Thus I agree it would be interesting to be able to run the tests via a JVM inside docker but that feels like secondary to this PR's main usecase of optionally allow running native agent?

vsevel commented:
> but that feels like secondary to this PR's main usecase of optionally allow running native agent?

I agree this is secondary. this PR brings a lot of value, even without the capability to not rely on a local installation of graalvm. it is not part of the mvp.
for us it is the difference between getting it to run every time for every app, and from time to time when we have an issue with the app, or the very first time we target native.
I doubt we will be able to convince the soft factory team to follow up with installing the various versions of graalvm we need (today we do everything in docker, including maven), but that may be different in other organizations.
another option for us is to convince developers to install graalvm on their dev machine. but they will do so only if they really need to (even if I write a procedure). the fact that we are on windows, adds the extra requirement that we have to install the vs tools. 
so all I am saying is that requiring a local graalvm install will limit the number of times we are using this feature (whether or not we look at the build result).
my initial thought was that I would add this profile to our default quarkus build pipeline. I won't be able to do that. but like I said, there is already a lot of value, even without docker support.

galderz committed:
Enable option to run integration tests with native image agent

* Added a build step to generate native image agent filters.
These limit the generated configuration to avoid some of
configuration that Quarkus already takes care of.
* Native image agent goal added to Quarkus maven plugin
to post-process native image agent configuration files.
In particular, the native image agent filters do not
impact resource configuration, so these files are
post-processed to remove configuration that Quarkus takes care of.
* Added a JSON parser to read the resources configuration
that the native image agent generates.
* Added the ability to transform JSON, which links the
JSON parser and JSON emitter.
* Added native agent integration test.
* The integration tests that run with native image agent
execute Quarkus JVM mode using the `java` within the Mandrel
builder image.
* Refactored container code to enable container launcher
to share container runtime parameters common to other similar
use cases. In particular, volume related options need to be
carefully constructed.

galderz commented on core/builder/src/main/java/io/quarkus/builder/Json.java:
```diff
@@ -311,25 +409,67 @@ protected JsonObjectBuilder self() {
+                final String attribute = member.attribute().value();
+                final JsonReader.JsonValue value = member.value();
+                if (value instanceof JsonReader.JsonString) {
+                    put(attribute, ((JsonReader.JsonString) value).value());
```
I looked into this and I'm not sure about it. The object builder methods `put` calls take an attribute and value there, whereas the array builder ones only take a value, so I don't think it's a good idea to bunch them all together at object builder class. The duplication could be abstracted away in other ways but I'm unsure there's a real tangible benefit since the JSON grammar is set and we're not going to evolve this further.

galderz commented:
I have pushed an update to rework a few aspects and remove baggage from previous experiments. Here are the highlights:

* Generating native image configuration is produced by JVM mode failsafe integration tests instead of surefire unit tests. Running JVM mode integration tests is generally achieved with `mvn verify -DskipITs=false` command.
* To get JVM mode integration tests running with the native image agent, append `-Dquarkus.test.integration-test-profile=test-with-native-agent` to the invocation in the previous point.
* When `-Dquarkus.test.integration-test-profile=test-with-native-agent` is passed in, the Quarkus process for the JVM mode integration tests runs via the Mandrel builder container image `java` process. E.g. `podman run --name quarkus-integration-test-ioSIv -i --rm --user 501:20 -p 8081:8081 -p 8444:8444 --entrypoint java -v /Users/galder/1/mendrugo/metadata/new-project/target:/project --env QUARKUS_LOG_CATEGORY__IO_QUARKUS__LEVEL=INFO --env QUARKUS_HTTP_PORT=8081 --env QUARKUS_HTTP_SSL_PORT=8444 --env TEST_URL=http://localhost:8081 --env QUARKUS_PROFILE=test-with-native-agent --env QUARKUS_TEST_INTEGRATION_TEST_PROFILE=test-with-native-agent quay.io/quarkus/ubi-quarkus-mandrel-builder-image:jdk-21 -agentlib:native-image-agent=access-filter-file=quarkus-access-filter.json,caller-filter-file=quarkus-caller-filter.json,config-output-dir=native-image-agent-base-config, -jar quarkus-app/quarkus-run.jar`.
* Switching to running JVM integration tests in-container as shown above is done within the `DefaultDockerContainerLauncher*` classes.
* A new Maven profile is no longer needed. Existing Maven projects need to add the `<goal>native-image-agent</goal>` to the `quarkus-maven-plugin` configuration. This goal takes care of the post-processing of JSON files. Generating native image agent filters have been bolted on to the `GenerateCodeTestsMojo` mojo, which gets activated when `test-with-native-agent` test profile is passed in .
* Running through failsafe simplifies the filter package definition, only need a few hardcoded high level packages as callers and/or filters, e.g. `io.quarkus`, `org.jboss`...etc, because surefire unit test dependencies no longer need filtering out (e.g. groovy, maven...etc).

This update also addresses all PR comments so far. Pretty-printing of the JSON files has been removed since that can be executed post-build with `jq` or similar tools.

I have also updated the description of https://github.com/quarkusio/quarkus/issues/36822 to reflect how the current prototype works.

geoand commented:
Thanks a lot @galderz, this is very nice!

I'll have a closer look tomorrow (hopefully).

geoand commented:
At the very least, `QuarkusCodestartGenerationTest` will also need to be updated.

geoand commented:
Furhermore, the PR should definitely be rebased onto `main` and the commits probably cleaned up a bit

galderz commented:
Fixed the unit tests, rebased and squashed commits. Hold on integrating to hear what @zakkak and @vsevel think about the new iteration.

zakkak reviewed (approved):
LGTM, in general. Added a few comments/questions for potential further improvement.

galderz commented on devtools/maven/src/main/java/io/quarkus/maven/GenerateCodeTestsMojo.java:
```diff
@@ -1,16 +1,107 @@
+        result.add("java");
+        result.add("javax");
+        result.add("jdk");
+        result.add("sun");
```
I selected these specific packages based on the getting started Quarkus application, so one of the most basic Quarkus applications. Without these exclusions some JNI and reflection configuration was being added which seemed unnecessary. See the following attached files for examples:

[jni-origins.txt](https://github.com/quarkusio/quarkus/files/14943352/jni-origins.txt)
[reflect-origins.txt](https://github.com/quarkusio/quarkus/files/14943356/reflect-origins.txt)

These specific exclusions might be too restrictive for some other use cases, e.g. AWT/images, where reflection or library load calls come from inside the JDK. I didn't want to make these hardcoded exclusions very fine grained, so I think we can go 2 ways here:

a) Keep these wide exclusions and revisit them as needed as we try different use cases.
b) Remove these wide exclusions and accept some extra unnecessary configuration.

It might be safer to go with b) here.

zakkak replied:
I think we should do some tests to see what's the actual impact of b) on the image size, and if possible on runtime performance.

Since the agent detects that these are actually accessed I am curious how/why we can get away without them.

galderz replied:
> Sidenote: Since the agent detects that these are actually accessed I am curious how/why we can get away without them.

In some cases svm can do constant folding of those, e.g.
```
               └── io.quarkus.arc.impl.ArcContainerImpl#<init>(io.quarkus.arc.CurrentContextFactory,boolean)
                   └── java.util.concurrent.atomic.AtomicBoolean#<clinit>()
                       └── java.lang.invoke.MethodHandles$Lookup#findVarHandle(java.lang.Class,java.lang.String,java.lang.Class) - [ {   "name":"java.util.concurrent.atomic.AtomicBoolean",   "fields":[{"name":"value"}] } ]
s":[{"name":"value"}] } ]
```

SVM handles use cases like this by transforming the graph by doing

galderz replied:
Also, once you have something running and you understand the configuration, you can see if you want to trim back, say to speed up build times or reduce image size.

galderz replied:
I've removed JDK package filter exclusions and added some comments. The extra configuration added for the getting started Quarkus app is quite small so it's a reasonable thing to do.

galderz commented on devtools/maven/src/main/java/io/quarkus/maven/NativeImageAgentMojo.java:
```diff
@@ -0,0 +1,122 @@
+    private final Pattern resourceSkipPattern;
+
+    public NativeImageAgentMojo() {
+        resourceSkipPattern = discardPattern("application.properties", "java[.\\/](lang|net|nio|time)", "jakarta", "jboss",
```
If we go with b) above, it would make sense to remove `"java[.\\/](lang|net|nio|time)"` to keep things symmetric. All other resources are related to resources that are common to all Quarkus applications.

galderz commented on integration-tests/maven/src/test/java/io/quarkus/maven/it/NativeAgentIT.java:
```diff
@@ -0,0 +1,33 @@
+        assertThat(runJvmITsWithAgent.getProcess().waitFor()).isZero();
+
+        MavenProcessInvocationResult runNativeITs = running
+                .execute(List.of("verify", "-Dnative", "-Dquarkus.native.agent-configuration-apply"), Map.of());
```
See https://gist.github.com/galderz/34785597c77893b137f2e47fad1de394

zakkak replied:
Good, thanks.

I think it could be handy to print the mvn commands so that on failures we can copy paste them to reproduce and inspect the failure.

galderz replied:
Couldn't see an easy way to achieve just this, but instead I added some `INFO` logging to Maven process invoker so that it prints out the command being executed.

zakkak commented:
Still LGTM. Thanks

geoand commented:
Once all the GraalVM specifics have been addressed, I want to take a final look and see if I've missed anything Quarkus specific here.

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/NativeImageAgentConfigDirectoryBuildItem.java:
```diff
@@ -0,0 +1,15 @@
+
+import io.quarkus.builder.item.SimpleBuildItem;
+
+public final class NativeImageAgentConfigDirectoryBuildItem extends SimpleBuildItem {
```
It would be nice to have some Javadoc about what this does

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/steps/ApplyNativeImageAgentConfigStep.java:
```diff
@@ -0,0 +1,50 @@
+    private static final Logger log = Logger.getLogger(ApplyNativeImageAgentConfigStep.class);
+
+    @BuildStep(onlyIf = NativeOrNativeSourcesBuild.class)
+    void transformConfig(NativeConfig nativeConfig,
```
A quick note in Javadoc about what this does would be nice

geoand commented on test-framework/junit5/src/main/java/io/quarkus/test/junit/launcher/ArtifactLauncherProvider.java:
```diff
@@ -11,7 +11,7 @@ public interface ArtifactLauncherProvider {
      *
      */
-    boolean supportsArtifactType(String type);
+    boolean supportsArtifactType(String type, String testProfile);
```
This is a breaking change, although it should not really affect anyone

galderz replied:
Hmmmm, yeah. I can look into trying to make the change in a more backwards compatible way?

geoand replied:
Nah, it's extremely unlikely anyone but us is using it and even if they are, this was never documented or communicated as part of the API

geoand commented on test-framework/junit5/src/main/java/io/quarkus/test/junit/launcher/DockerContainerLauncherProvider.java:
```diff
@@ -42,25 +47,73 @@ public DockerContainerArtifactLauncher create(CreateContext context) {
+            launcherInit(context, launcher, config, containerImage, pullRequired, Optional.empty(), volumeMounts(config),
+                    Collections.emptyList());
             return launcher;
         } else {
```
I would like have a general comment here describing on a high level what this block is  doing

geoand commented on test-framework/maven/src/main/java/io/quarkus/maven/it/verifier/MavenProcessInvoker.java:
```diff
@@ -56,6 +56,9 @@ public InvocationResult execute(InvocationRequest request) throws MavenInvocatio
         try {
             cli = cliBuilder.build(request);
+            if (logger != null) {
+                logger.info("Running Maven CLI command: " + cli);
```
Do we really want this to be INFO?

galderz replied:
If CI shows an error from one of these tests, you would right away have the info on what was executed. As opposed to making it DEBUG, and then having to run this with debug locally to see what got executed. I'm fine either way TBH. @zakkak wanted to see this logged, WDYT?

zakkak replied:
I would prefer it to stay at INFO level, but DEBUT would be OK as well.

galderz replied:
I've switched it to DEBUG.

geoand reviewed (commented):
Thanks a lot for this!

I added some minor comments which once addressed can merge

galderz commented:
Re: I added some minor comments which once addressed can merge

Before merging I'd like to get some final feedback from @vsevel to see how things look on his side.

galderz commented:
@geoand I've added several javadoc comments, please have a look.

geoand commented:
@galderz should we merge this or do want to wait for more feedback from @vsevel ?

galderz commented:
@geoand See https://github.com/quarkusio/quarkus/issues/36822#issuecomment-2074190871 - I'm looking into the issue @vsevel reported there.

galderz commented:
The windows issue has now been fixed. Checking if @vsevel has additional comments to make. I'll squash commits in the mean time.

galderz commented:
@geoand I've squashed commits. It looks good for @vsevel too. Documentation needs adding, I can either add that as part o this PR or a follow up PR, thoughts?

galderz commented:
FYI see #40360 and #40361 for follow up enhancements.

geoand commented:
> @geoand I've squashed commits. It looks good for @vsevel too. Documentation needs adding, I can either add that as part o this PR or a follow up PR, thoughts?

Cool!

Sorry for missing the ping.
Can you please add a little bit in the docs and rebase it so we can merge after?

Thanks

galderz commented:
I've updated the PR with the latest main changes and I've added documentation.

aloubyansky commented on docs/src/main/asciidoc/native-reference.adoc:
```diff
@@ -528,6 +528,283 @@ ${FG_HOME}/stackcollapse.pl < out.stacks | ${FG_HOME}/flamegraph.pl \
+2024-05-14 16:29:53,941 INFO  [io.qua.tes.com.DefaultDockerContainerLauncher] (main) Executing "podman run --name quarkus-integration-test-PodgW -i --rm --user 501:20 -p 8081:8081 -p 8444:8444 --entrypoint java -v /tmp/new-project/target:/project --env QUARKUS_LOG_CATEGORY__IO_QUARKUS__LEVEL=INFO --env QUARKUS_HTTP_PORT=8081 --env QUARKUS_HTTP_SSL_PORT=8444 --env TEST_URL=http://localhost:8081 -- ...
+...
+[INFO]
+[INFO] --- quarkus:999-SNAPSHOT:native-image-agent (default) @ new-project ---
```
```suggestion
[INFO] --- quarkus:{quarkus-version}:native-image-agent (default) @ new-project ---
```

aloubyansky commented on docs/src/main/asciidoc/native-reference.adoc:
```diff
@@ -528,6 +528,283 @@ ${FG_HOME}/stackcollapse.pl < out.stacks | ${FG_HOME}/flamegraph.pl \
+----
+$ ./mvnw verify -DskipITs=false -Dquarkus.test.integration-test-profile=test-with-native-agent
+...
+[INFO] --- quarkus:999-SNAPSHOT:native-image-agent (default) @ new-project ---
```
```suggestion
[INFO] --- quarkus:{quarkus-version}:native-image-agent (default) @ new-project ---
```

aloubyansky commented on docs/src/main/asciidoc/native-reference.adoc:
```diff
@@ -528,6 +528,283 @@ ${FG_HOME}/stackcollapse.pl < out.stacks | ${FG_HOME}/flamegraph.pl \
+----
+$ ./mvnw verify -Dnative -Dquarkus.native.agent-configuration-apply
+...
+[INFO] --- quarkus:999-SNAPSHOT:build (default) @ new-project ---
```
```suggestion
[INFO] --- quarkus:{quarkus-version}:build (default) @ new-project ---
```

aloubyansky commented on docs/src/main/asciidoc/native-reference.adoc:
```diff
@@ -528,6 +528,283 @@ ${FG_HOME}/stackcollapse.pl < out.stacks | ${FG_HOME}/flamegraph.pl \
+----
+$ ./mvnw -X verify -DskipITs=false -Dquarkus.test.integration-test-profile=test-with-native-agent
+...
+[INFO] --- quarkus:999-SNAPSHOT:native-image-agent (default) @ new-project ---
```
```suggestion
[INFO] --- quarkus:{quarkus-version}:native-image-agent (default) @ new-project ---
```

aloubyansky commented on devtools/maven/src/main/java/io/quarkus/maven/GenerateCodeTestsMojo.java:
```diff
@@ -1,16 +1,100 @@
         generateCode(getParentDirs(mavenProject().getTestCompileSourceRoots()),
                 path -> mavenProject().addTestCompileSourceRoot(path.toString()), true);
+
+        final String testProfile = System.getProperty("quarkus.test.integration-test-profile");
```
Not a blocker but this can be approved by consulting other config sources.
Maven project properties would be a trivial one to add but others could be made to work as well.

galderz replied:
Fair point, this is a leftover from my first experiments here. I'll see if I can hook another source instead.

aloubyansky commented on devtools/maven/src/main/java/io/quarkus/maven/NativeImageAgentMojo.java:
```diff
@@ -0,0 +1,123 @@
+import io.quarkus.builder.json.JsonValue;
+
+@Mojo(name = "native-image-agent", defaultPhase = LifecyclePhase.POST_INTEGRATION_TEST, requiresDependencyResolution = ResolutionScope.COMPILE_PLUS_RUNTIME, threadSafe = true)
+public class NativeImageAgentMojo extends QuarkusBootstrapMojo {
```
@galderz could you please add a class-level Javadoc here? It will be displayed in the description of this goal by Maven then.

galderz committed:
Update docs/src/main/asciidoc/native-reference.adoc

Co-authored-by: Alexey Loubyansky <olubyans@redhat.com>

galderz committed:
Update docs/src/main/asciidoc/native-reference.adoc

Co-authored-by: Alexey Loubyansky <olubyans@redhat.com>

galderz committed:
Update docs/src/main/asciidoc/native-reference.adoc

Co-authored-by: Alexey Loubyansky <olubyans@redhat.com>

galderz committed:
Update docs/src/main/asciidoc/native-reference.adoc

Co-authored-by: Alexey Loubyansky <olubyans@redhat.com>

galderz committed:
Add javadoc to mojo

galderz committed:
Check maven and system properties when looking for test profile

galderz commented:
@aloubyansky Pushed commits to fix your comments.

aloubyansky reviewed (approved):
Thanks!

vsevel commented:
awesome @galderz

vsevel commented:
just tried out in juste released `3.12`
works like a charm @galderz 
thanks and congrats!
